router/data_store.go

```diff
@@ -39,196 +33,206 @@ type SyncHandler interface {
 	}
-	_, err = s.etcd.Create(s.path(r.ID), string(data), 0)
-	if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 105 {
-		err = ErrExists
+	if r.UpdatedAt == nil {
+		r.UpdatedAt = new(time.Time)
 	}
-	return err
+	return d.pgx.QueryRow(sqlAddRoute, r.ParentRef, r.Type, r.Config).Scan(&r.ID, r.CreatedAt, r.UpdatedAt)
 }
 
-func (s *etcdDataStore) Set(r *router.Route) error {
-	data, err := json.Marshal(r)
-	if err != nil {
-		return err
+var sqlSetRoute = `UPDATE routes SET parent_ref = $1, type = $2, config = $3::json WHERE route_id = $4 RETURNING updated_at`
+
+func (d *pgDataStore) Set(r *router.Route) error {
+	if r.UpdatedAt == nil {
+		r.UpdatedAt = new(time.Time)
+	}
+	err := d.pgx.QueryRow(sqlSetRoute, r.ParentRef, r.Type, r.Config, r.ID).Scan(r.UpdatedAt)
+	if err == pgx.ErrNoRows {
+		return ErrNotFound
 	}
-	_, err = s.etcd.Set(s.path(r.ID), string(data), 0)
 	return err
 }
 
-func (s *etcdDataStore) Remove(id string) error {
-	_, err := s.etcd.Delete(s.path(id), true)
-	if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 100 {
-		return ErrNotFound
-	}
+var sqlRemoveRoute = `UPDATE routes SET deleted_at = now() WHERE route_id = $1`
+
+func (d *pgDataStore) Remove(id string) error {
+	_, err := d.pgx.Exec(sqlRemoveRoute, id)
 	return err
 }
 
-func (s *etcdDataStore) Get(id string) (*router.Route, error) {
-	res, err := s.etcd.Get(s.path(id), false, false)
+var sqlGetRoute = `SELECT parent_ref, type, config, created_at, updated_at FROM routes WHERE route_id = $1 AND type = $2 AND deleted_at IS NULL`
+
+func (d *pgDataStore) Get(id string) (*router.Route, error) {
+	r := &router.Route{ID: id, Config: new(router.Config), CreatedAt: new(time.Time), UpdatedAt: new(time.Time)}
+	row := d.pgx.QueryRow(sqlGetRoute, id, d.rt)
+
+	err := row.Scan(&r.ParentRef, &r.Type, r.Config, r.CreatedAt, r.UpdatedAt)
+	if err == pgx.ErrNoRows {
+		return nil, ErrNotFound
+	}
 	if err != nil {
-		if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 100 {
-			err = ErrNotFound
-		}
 		return nil, err
 	}
-	r := &router.Route{}
-	err = json.Unmarshal([]byte(res.Node.Value), r)
-	return r, err
+
+	return r, nil
 }
 
-func (s *etcdDataStore) List() ([]*router.Route, error) {
-	res, err := s.etcd.Get(s.prefix, false, true)
+var sqlListRoutes = `SELECT route_id, parent_ref, type, config, created_at, updated_at FROM routes WHERE type = $1 AND deleted_at IS NULL`
+
+func (d *pgDataStore) List() ([]*router.Route, error) {
+	rows, err := d.pgx.Query(sqlListRoutes, d.rt)
 	if err != nil {
-		if e, ok := err.(*etcd.EtcdError); ok && e.ErrorCode == 100 {
-			err = nil
-		}
 		return nil, err
 	}
 
-	routes := make([]*router.Route, len(res.Node.Nodes))
-	for i, node := range res.Node.Nodes {
-		r := &router.Route{}
-		if err := json.Unmarshal([]byte(node.Value), r); err != nil {
+	routes := []*router.Route{}
+	for rows.Next() {
+		r := &router.Route{Config: new(router.Config), CreatedAt: new(time.Time), UpdatedAt: new(time.Time)}
```

Comment:
This route initialization seems to happen a lot (I assume to work around pgx's scanner). Perhaps make a helper?